plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java

```diff
@@ -168,11 +175,27 @@ public void performAction(IncrementalTaskInputs inputs) throws Exception {
 				.steps(steps)
 				.exceptionPolicy(exceptionPolicy)
 				.build()) {
+			// determine if a list of files has been passed in
+			Predicate<File> shouldInclude;
+			Project project = getProject();
+			if (project.hasProperty(FILES_PROPERTY) && project.property(FILES_PROPERTY) instanceof String) {
+				Object rawIncludePatterns = project.property(FILES_PROPERTY);
+				final String[] includePatterns = ((String) rawIncludePatterns).split(",");
+				final List<Pattern> compiledIncludePatterns = Arrays.stream(includePatterns)
+						.map(Pattern::compile)
+						.collect(Collectors.toList());
+				shouldInclude = file -> compiledIncludePatterns
+						.stream()
+						.anyMatch(filePattern -> filePattern.matcher(file.getAbsolutePath())
+								.matches());
+			} else {
+				shouldInclude = file -> true;
+			}
```

Comment:
This commit introduces a subtle bug in up-to-date checking.  The gist of it is that gradle up-to-date checking is based on `f(inputDir) => outputDir`, but with Spotless it's `f(inputDir) => inputDir`, which means we need a hack to make things work, and this commit breaks the hack.  You can read about it in full detail [here](https://github.com/diffplug/spotless/issues/144#issuecomment-331615009).

The fix is that you need to read the `-PspotlessFiles` property at configuration time, when [the tasks are being created](https://github.com/diffplug/spotless/blob/d57e26d9a90c67050c0080b59aa28facdb7c9908/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java#L70-L71), and set `String spotlessFiles` as an `@Input` property of `SpotlessTask`, [like this](https://github.com/diffplug/spotless/blob/d57e26d9a90c67050c0080b59aa28facdb7c9908/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java#L50-L60).  When you make your `shouldInclude` predicate above, read from the `spotlessFiles` field rather than the property.

Reply:
updated